### ISSUE TITLE
Binary encoding of value type codes with heap type indices

### ISSUE THREAD

**Issue by titzer** [WebAssembly/gc#337]
We haven't chosen the final binary encoding for value type codes introduced in this proposal.

One issue is that decoding value types is a very hot operation in in-place interpretation, particularly the declaration of local variables as well as block types (though block types are just decoded and ignored). Until this proposal, all value types are decodable as a single byte (or single LEB--we've been inconsistent on that so far). With this proposal, some value type codes have a heaptype immediate. Thus an in-place interpreter can no longer simply decode (or skip) a single LEB.

I've resorted to doing a table lookup to see if the value type code indicates that a heaptype index should follow (`ref` or `ref null`). But with one tweak to the assignment of type codes, it's possible to turn that into a bit-test.

```
11111111 I32 = -1
11111110 I64 = -2
11111101 F32 = -3
11111100 F64 = -4
11111011 V128 = -5
11110000 FUNCREF = -16
11101111 EXTERNREF = -17
11101110 ANYREF = -18
11101101 EQREF = -19
11101100 REF_NULL = -20
11101011 REF = -21
11101010 I31REF = -22
11101001 NULLFUNCREF = -23
11101000 NULLEXTERNREF = -24
11100111 STRUCTREF = -25
11100110 ARRAYREF = -26
11100101 NULLREF = -27
11000000 EmptyBlock = -64
```

It turns out that we have encoding space to make the 6th bit the determinant as to whether a heaptype immediate follows. We just need to make `ref` and `ref null` < `-32`. That makes the 6th bit `0` in case of `ref`, `ref null`; empty block still needs a separate check. (Perhaps it is still possible to rearrange so that only one test is necessary).

That constrains the "short encoding" space to 32 entries if we want to preserve the bit test property in future additions.

**Comment by rossberg** [comment]
Don't forget the codes for structured types (func, struct, array), which already occupy the range -32..-34. They are supposed to share a single code space with value types, since we may have a future need to allow them in the same place, e.g., consider nested structs.

Also, it's not just the number of code space entries to consider. This optimisation is based on the implicit assumption that no other immediates will ever be added to value types. I highly doubt that is a valid assumption, it is rather easy to predict new immediates like attribute bytes etc. And we don't have enough space in a byte to afford a bit each time, so this scheme would not scale.

**Comment by titzer** [comment]
> Don't forget the codes for structured types (func, struct, array), which already occupy the range -32..-34. They are supposed to share a single code space with value types, since we may have a future need to allow them in the same place, e.g., consider nested structs.

I think we should probably escape those ones in particular in nested structs, since i don't think we them to appear other places that a value type can occur, like a local. Fields and array elements are already classified as storage types, so I don't see an issue; an interpreter doesn't look at those. Storage types have a mutability byte, which can serve as a flags byte in the future.

> And we don't have enough space in a byte to afford a bit each time, so this scheme would not scale.

The bit only has to distinguish between one-byte value types and multi-byte value types.

**Comment by rossberg** [comment]
> I think we should probably escape those ones in particular in nested structs

Yes, that would work.

> The bit only has to distinguish between one-byte value types and multi-byte value types.

I don't feel strongly, but we haven't done any such nano-optimisations of opcode bit patterns elsewhere in Wasm, so it feels a bit over the top. What do other folks think?

**Comment by tlively** [comment]
I don't feel strongly either way. I would be fine optimizing the numbering now, but would be unhappy if in the future we had to consider whether potential extensions would be compatible with the numbering optimization scheme we end up with.

**Comment by titzer** [comment]
@tlively That's fair. I agree it's a nano-optimization. At this juncture it doesn't cost us anything, and I'd be fine to give up on it the minute it makes a design problem harder in the future.

**Comment by rossberg** [comment]
Coming back to this, introducing overlap between different type constructor spaces has a potential price that is easily higher than what's to be gained from this nano-optimisation, the possibility of escaping notwithstanding. I think we should refrain from doing that if we can help it.

That said, I think we can enable the same optimisation without relying on bit fiddling: if we sort the nullary constructors first and then the unary ones (ignoring the special void case), then you can simply compare with the lowest non-nullary code for the same effect.

So I would propose e.g. (the * ones we cannot change):
```
*I32 = -1
*I64 = -2
*F32 = -3
*F64 = -4
*V128 = -5
*FUNCREF = -16
*EXTERNREF = -17
ANYREF = -18
EQREF = -19
I31REF = -20
STRUCTREF = -21
ARRAYREF = -22
NULLREF = -23
NULLFUNCREF = -24
NULLEXTERNREF = -25
REF_NULL = -28
REF = -29
*FUNC = -32
STRUCT = -33
ARRAY = -34
*EmptyBlock = -64
```
This way, you just compare opcode < -28 to test if it has immediate(s). FUNC, STRUCT, ARRAY actually fit nicely in this pattern.

(We could also move some of the reference type constructors before FUNCREF if we want to make more space for non-nullaries.)

**Comment by titzer** [comment]
Ok, I have to think about if that comparison works with a 1-byte LEB.

**Comment by titzer** [comment]
Thought about this, and this is ordering is fine. AFAICT, it will require two comparisons (and will always) because it requires an upper-bit check for an extended LEB and then a comparison (unsigned >= if 1-byte) against the nullary range.

**Comment by tlively** [comment]
Is there anything left to decide here? What needs to be done to close this?

**Comment by rossberg** [comment]
This would be closed by #372.
